### DISCUSSION TITLE
Do scheduled backups use a write lock?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#4422] [Q&A]
I'm tracking down the source of hourly instability on PocketHost and noticed that it coincides with the hourly scheduled backups on the primary database (mothership). I'm still investigating, but I see some log patterns that might indicate `update` API calls fail at the top of each hour.

So I just thought I'd ask, does PocketBase use a write lock when it is performing a backup?

Edit: Now I see this note:

> Note that the application will be temporary set in read-only mode during the backup's ZIP generation.

So I guess that might be the trouble? Are API writes queued, or do they return an error during this time?

**Comment by ganigeorgiev** [comment]
Yes, the backups starts a write transaction which effectively blocks all write requests because we allow only 1 write/transaction connection at a time (at least when PocketBase is the only one connected to the db file).

We don't throw an error when there is a write during a backup.
Other write requests will have to wait first the backup to complete before proceeding further.

**Reply by benallfree** [reply]
Do you have any thoughts or plans to do something incremental? I experimented with something a while ago and I think SQLite does have a way to do paginated incremental backups. 

Otherwise I'll take a look at Litestream or something as an option.

edit: Online Backup API https://www.sqlite.org/backup.html

**Reply by ganigeorgiev** [reply]
I'm not sure. I haven't researched in details the available options (creating an incremental backup for the local uploaded files is the more difficult part as we will have to reimplement something like rclone).

**Reply by benallfree** [reply]
I dug up my old code that PocketHost used to perform online backups before the PocketBase admin had a backup tool.

https://github.com/pockethost/pockethost/blob/a8ed9b9c933575bf2302bf8d5c2049b386ce794c/packages/daemon/src/util/backupInstance.ts#L53

Basically calling `sqlite_backup_step(5)` over and over until the state turned to finished. https://www.sqlite.org/c3ref/backup_finish.html#sqlite3backupstep

**Reply by ganigeorgiev** [reply]
But this is only for the database. When using the local filesystem for storing uploaded files (although I'm not sure if this matter in your case) the backup needs to block anyway to ensure consistency with the db records.

I'll consider changing the backup behavior in the future and/or maybe provide an option to skip the lock for the cases when the local filesystem is not used or it is not that critical if the `storage` directory is not synced 1:1 with the db file.

**Reply by benallfree** [reply]
The backup includes `storage`? I assume only if it's local (not s3)?

**Reply by ganigeorgiev** [reply]
Yes, when the local filesystem is used for storing uploaded files the backup includes the `storage` directory (for S3 I wasn't able to find an efficient way to perform the backup and this is left to the users to tailor it based on their used provider).

**Reply by benallfree** [reply]
I was experimenting with doing an s3 fuse mount using `rclone mount` for `storage` and `backups`. Some users have 30GB+ in `storage` and I'm looking for a way to offload it without requiring them to migrate or use their own s3 key. It sounds like any backup they initiated would cause all the files to be pulled local and zipped, is that right?

Is there any way to prevent PocketBase from including `storage` in backups, other than using their own s3 key?

**Reply by ganigeorgiev** [reply]
> It sounds like any backup they initiated would cause all the files to be pulled local and zipped, is that right?

I haven't tested it to confirm but if it is treated as regular directory then Yes I expect it to be included in the backup zip as long as `fs.WalkDir` can traverse it.

> Is there any way to prevent PocketBase from including storage in backups, other than using their own s3 key?

No, at the moment we don't have such option but I'm not sure if that will be good idea in general. PocketBase backups are usually intended for a full "snapshot" of the application state and the fact that we don't backup S3 files is more of a limitation of the current implementation and not a wanted effect because the restore will be incomplete.
